map/src/main/kotlin/edu/artic/map/MapFragment.kt

```diff
@@ -515,17 +554,95 @@ class MapFragment : BaseViewModelFragment<MapViewModel>() {
         viewModel
                 .centerFullObjectMarker
                 .subscribe { nid ->
-                    fullObjectMarkers.firstOrNull { marker ->
-                        val tag = marker.tag
-                        tag is ArticObject && tag.nid == nid
-                    }?.let { marker ->
-                        val currentZoomLevel = map.cameraPosition.zoom
-                        map.animateCamera(CameraUpdateFactory.newLatLngZoom(marker.position, Math.max(ZOOM_LEVEL_THREE, currentZoomLevel)))
+                    fullObjectsLock.withLock {
+                        fullObjectMarkers.firstOrNull { (model, _) ->
+                            model is MapItem.Object && model.item.nid == nid
+                                    ||
+                                    model is MapItem.TourIntro && model.item.nid == nid
+                        }?.let { (_, marker) ->
+                            val currentZoomLevel = map.cameraPosition.zoom
+                            map.animateCamera(CameraUpdateFactory.newLatLngZoom(marker.position, Math.max(ZOOM_LEVEL_THREE, currentZoomLevel)))
+                        }
                     }
 
                 }.disposedBy(disposeBag)
     }
 
+    @UiThread
+    fun bindToMap(mapItem: MapItem<*>, mapMode: MapViewModel.DisplayMode) {
+        when (mapItem) {
+            is MapItem.Annotation -> {
+                val annotation = mapItem.item
+                when (annotation.annotationType) {
+                    ArticMapAnnotationType.DEPARTMENT -> {
+                        loadDepartment(mapItem)
+                    }
+                }
+            }
+            is MapItem.Gallery -> {
+                loadGalleryNumber(mapItem)
+            }
+            is MapItem.Object -> {
+                loadObject(mapItem, mapMode)
+            }
+            is MapItem.TourIntro -> {
+                loadTourObject(mapItem, mapMode)
+            }
+
+        }
+    }
+
+
+    @WorkerThread
+    fun bindMarkersAsynchronously(itemList: List<MapItem<*>>, mapMode: MapViewModel.DisplayMode) {
+        val coreCount = Runtime.getRuntime().availableProcessors()
+
+        val batchedItems : List<List<MapItem<*>>>
+        batchedItems = if (requireContext().isResourceConstrained()) {
+            // We don't have that many resources. Only pass along 2 x 'number of cores' items at a time
+            itemList.chunked(2 * coreCount)
+        } else {
+            // Work with 3 x 'number of processor cores' items at a time.
+            itemList.chunked(3 * coreCount)
+        }
+
+                Observables.zip(
+                        // There's an Observable operator 'window', but that's (severely) restricted to a single subscriber
+                        batchedItems.toObservable(),
+                        // The 'interval' here will make sure only one event is emitted every 100 ms
```

Comment:
just use `throttleLast()` with 100 MS. does it more cleanly.

Reply:
throttle last will also emit if theres more events within the 100 ms timeframe. this solution really just delays it by 100ms.

Reply:
http://reactivex.io/documentation/operators/sample.html

Reply:
```
 Returns an Observable that emits only the last item emitted by the source Observable during sequential
     * time windows of a specified duration.
     * <p>
     * This differs from {@link #throttleFirst} in that this ticks along at a scheduled interval whereas
     * {@link #throttleFirst} does not tick, it just tracks passage of time.
```
This might work so it continues

Reply:
and increasing the delay to like to 500 ms might help. it will be slightly delayed but will pick up anything that happens within the 500ms and use the last even before emitting

Reply:
(we talked a little offline about the intent of this code)

The main goal is not so much to prevent too many events from being handled, but to slow down the rate at which we handle them. So `window` was very nearly what I needed. @agrosner communicated that he is working on a great improvement to the way `itemList` is passed in, so with an eye towards replacing it soon the code may stand.